src/main/kotlin/no/nav/tilleggsstonader/sak/vilkår/stønadsperiode/StønadsperiodeService.kt

```diff
@@ -1,38 +1,105 @@
 package no.nav.tilleggsstonader.sak.vilkår.stønadsperiode
 
+import no.nav.tilleggsstonader.sak.behandling.BehandlingService
+import no.nav.tilleggsstonader.sak.infrastruktur.exception.feilHvis
 import no.nav.tilleggsstonader.sak.vilkår.stønadsperiode.domain.Stønadsperiode
 import no.nav.tilleggsstonader.sak.vilkår.stønadsperiode.domain.StønadsperiodeRepository
 import no.nav.tilleggsstonader.sak.vilkår.stønadsperiode.dto.StønadsperiodeDto
 import no.nav.tilleggsstonader.sak.vilkår.stønadsperiode.dto.tilSortertDto
 import no.nav.tilleggsstonader.sak.vilkår.vilkårperiode.VilkårperiodeService
 import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
 import java.util.UUID
 
 @Service
 class StønadsperiodeService(
+    private val behandlingService: BehandlingService,
     private val stønadsperiodeRepository: StønadsperiodeRepository,
     private val vilkårperiodeService: VilkårperiodeService,
 ) {
     fun hentStønadsperioder(behandlingId: UUID): List<StønadsperiodeDto> {
         return stønadsperiodeRepository.findAllByBehandlingId(behandlingId).tilSortertDto()
     }
 
+    @Transactional
     fun lagreStønadsperioder(behandlingId: UUID, stønadsperioder: List<StønadsperiodeDto>): List<StønadsperiodeDto> {
-        // TODO valider behandling er ikke låst
+        feilHvis(behandlingService.hentBehandling(behandlingId).status.behandlingErLåstForVidereRedigering()) {
+            "Kan ikke lagre stønadsperioder når behandlingen er låst"
+        }
         validerStønadsperioder(behandlingId, stønadsperioder)
-        stønadsperiodeRepository.deleteAll() // TODO skal vi finne og oppdatere de som har blitt oppdatert/slettet
-        return stønadsperiodeRepository.insertAll(
-            stønadsperioder.map {
-                Stønadsperiode(
-                    id = it.id ?: UUID.randomUUID(),
-                    behandlingId = behandlingId,
+        return lagreOgOppdaterStønadsperioder(behandlingId, stønadsperioder)
+    }
+
+    private fun lagreOgOppdaterStønadsperioder(
+        behandlingId: UUID,
+        stønadsperioder: List<StønadsperiodeDto>,
+    ): List<StønadsperiodeDto> {
+        val tidligereStønadsperioder = stønadsperiodeRepository.findAllByBehandlingId(behandlingId)
+
+        slettStønadsperioder(tidligereStønadsperioder, stønadsperioder)
+
+        return oppdaterOgLagreNyeStønadsperioder(behandlingId, tidligereStønadsperioder, stønadsperioder)
```

Comment:
Sendes ikke også de slettede videre nå? De vil jo tas i mot av den at id ikke finnes, men virker fortsatt litt feil

Reply:
Siste del vet jeg ikke om var riktig, så stryker den. 

`tidligerePerioder` brukes videre selv om det er en liste som inneholder ting som nå er slettet. Er det dumt?

Reply:
Sender med alle tidligere perioder til alle 3 type håndtering av nye perioder fordi de har ulikt behov. 
Hva tenker du kan være dumt?

Reply:
Vet ikke helt. Føler det kan skje noe galt dersom den ikke stemmer med det som faktisk er lagret lenger, vet ikke helt hva men vet heller ikke hvorfor periodene som er slettet trenger å "være med videre"

Reply:
Tenker det er ryddig å splitte opp det som slett/oppdater/nye og at slett ikke returnerer bestående perioder, der blir ulogisk for meg.
Du får gjerne lage en commit i en branch og gi forslag :)

Reply:
Ja blir mye henting av tidligere perioder, så med mindre noe kan gå veldig galt av at de blir med videre så er det nok good